### ISSUE TITLE
[Feature Request] ePub pagination in reader

### ISSUE THREAD

**Issue by dathbe** [Kareadita/Kavita#738]
Is it possible to have the reader for epubs display text as a series of screen-size pages instead of a lengthy html page?  The bite-size pages that can be flipped with a click or a swipe are much more conducive to tablet reading than trying to scroll.

**Comment by majora2007** [comment]
I'm not sure this is possible. The the problem is that the page length is not determined by me nor is it something I can control. Essentially an ePub is an html document (webpage) so everything is self contained. So all styling is determined by the book and the html elements that make up the physical text. 

To provide what you're mentioning, you'd need to fix the amount of text on a "page", but how could I know where to split it or how the page is actually rendered? 

Because some epubs are comics which use absolute positioning, Kavita doesn't have any way to know any of this information. So I don't think this is a doable request.

**Comment by dathbe** [comment]
Calibre-Web and most other ereaders make it happen with standard epubs.  I couldn't tell you how they implement it exactly, but I know it's possible.  I can't get Calibre-Web to save my position between devices, so I was looking to Kavita as an alternative.  The only hangup so far is the reading experience (i.e., breaking up chapters into individual pages of blocks of text).  iPad screenshots attached of Kavita reading vs. Calibre-Web.  You'd obviously not want to do this for image files (cbr/cbz), but for epub/mobi/etc it would make for a more pleasant experience in my view.
![calibre-web](https://user-images.githubusercontent.com/14321825/141166905-c353d14b-c36c-42ce-86f8-0aa690e78aae.png)
![kavita](https://user-images.githubusercontent.com/14321825/141166922-6a32916e-1a72-4143-9a9b-431e984fab64.png)

**Comment by dathbe** [comment]
P.S. In the above images, it may not be obvious from looking at it, but in the Kavita app (dark mode) you have to scroll down to the bottom of the chapter, then next to the next chapter.  In the Calibre-Web app (light mode) you can just keep hitting the right side of the screen with your thumb to advance pages to the end of the book.

**Comment by majora2007** [comment]
In Kavita, we have what's called Tap to Paginate that you can enable in User settings or side nav. It lets you do what Calibre-web does, essentially tap the sides of the screen to paginate, like you'd do in our image reader. 

Are you saying the length of the document in Calibre-web is different from the length of the document in Kavita?

**Comment by dathbe** [comment]
So I think we may be using different definitions of "paginate".  I mean to break up the material into pages the way a paper book would have many pages to a chapter.  Kavita seems to display an entire chapter as a single "page".  I have "Tap to Paginate" enabled, but that just allows me to flip between chapters.  

I'm also not sure what you mean by "document" in your last sentence.  The book is exactly the same book.  But in the images above, the "page" or "document" that you are looking at is definitely a different length.  In Calibre-Web, the page is only what you see on the screen.  You cannot scroll down to continue reading, you must click to turn to the next page.  In Kavita, though, you have to scroll down to the end of the chapter, and only then can you click or swipe to turn the "page" to the next chapter.

In short, somehow Calibre-Web (and many other readers, though not Ubooquity) take the text and serve it up as bite-size pages exactly the right size to fit the screen.  That's what I'm hoping is relatively easy to implement.

**Comment by majora2007** [comment]
It's very interesting that in Calibre-web it behaves differently than Kavita. Kavita let's you move between pages in the book. It has no concept if the book has packed a whole chapter into one page or not. 

I will have to look into Calibre-web and see how they do it, but I wouldn't hold your breath for this functionality. Our reader works incredibly well and meets the spec. For example, we support javascript from running in the book, say if you have a technical book. I wonder how Calibre-web could handle that. 

I'll leave this open, but I'm not going to be looking at it anytime soon.

**Comment by majora2007** [comment]
Okay, I took a look at Calibre-web did it. I think it is doable, but it's a very PC oriented reading mode and from testing within Calibre-web, I found it to be broken more times than not (esp if the books are graphic novels). 

I'm not likely to implement this, but I will keep this open in case I feel adventurous to try an implementation that works across the board.

**Comment by majora2007** [comment]
Hello, if you want this still, please move it over with a good description (render ebook pages in 2 column layout like Caibre-web) to our feature request site: [https://feats.kavitareader.com/](https://feats.kavitareader.com/). All feature requests on Github are being closed out. You can reference this discussion in your request.
